File: src/Razor/src/Microsoft.AspNetCore.Razor.VSCode.Extension/src/SemanticTokensEdit.ts
/* --------------------------------------------------------------------------------------------
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 * ------------------------------------------------------------------------------------------ */

 export class SemanticTokensEdit {
    public readonly start: number;
    public readonly deleteCount: number;
    public readonly data?: Uint32Array;
    constructor(start: number, deleteCount: number, data?: Uint32Array) {
        this.start = start;
        this.deleteCount = deleteCount;
        this.data = data;
    }
}

File: src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Models/SemanticTokensEdit.cs
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

using System.Collections.Generic;
using System.Linq;
using Microsoft.Extensions.Internal;

namespace Microsoft.AspNetCore.Razor.LanguageServer.Semantic.Models
{
    internal class SemanticTokensEdit
    {
        public int Start { get; set; }
        public int DeleteCount { get; set; }
        public IEnumerable<uint> Data { get; set; }

        public override bool Equals(object obj)
        {
            if ((obj is null && this != null) || !(obj is SemanticTokensEdit other))
            {
                return false;
            }

            var equal = Start.Equals(other.Start);
            equal &= DeleteCount.Equals(other.DeleteCount);
            equal &= (Data is null && other.Data is null) || Enumerable.SequenceEqual(Data, other.Data);

            return equal;
        }

        public override int GetHashCode()
        {
            var combiner = HashCodeCombiner.Start();

            combiner.Add(Start);
            combiner.Add(DeleteCount);
            combiner.Add(Data);

            return combiner.CombinedHash;
        }
    }
}
